File: signature/python/check_webhook_signature.py
import base64
from hashlib import sha512

from ecdsa.util import sigdecode_der
from ecdsa import VerifyingKey

import typer


def check_signature(raw_webhook_body_path: str, signature_path: str, key_path: str):
    signature = open(signature_path).read()
    raw_webhook_body = open(raw_webhook_body_path).read()
    (
        timestamp,
        ubble_signature_version,
        ubble_signature,
    ) = signature.split(":")
    
    # First we create the signed_payload
    signed_payload = (f"{timestamp}:{raw_webhook_body}").encode("utf-8")
    
    # Use the public key from the dashboard (check version from ubble_signature_version)
    public_key_file = open(key_path)
    
    # Generate veryfing key signature
    raw_key = public_key_file.read()
    verifying_key = VerifyingKey.from_pem(raw_key)

    # Verify the signature
    result = verifying_key.verify(
        base64.b64decode(ubble_signature.encode("utf-8")),
        signed_payload,
        hashfunc=sha512,
        sigdecode=sigdecode_der,
        allow_truncate=True,
    )
    if result:
        print("Success")
    else:
        print("Could not validate signature")


if __name__ == "__main__":
    typer.run(check_signature)

File: signature/java/CheckWebhookSignature.java
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;

public class CheckWebhookSignature {

    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Usage: java CheckWebhookSignature <rawWebhookBodyPath> <signaturePath> <keyPath>");
            return;
        }

        String rawWebhookBodyPath = args[0];
        String signaturePath = args[1];
        String keyPath = args[2];

        try {
            String signature = Files.readString(Paths.get(signaturePath));
            String rawWebhookBody = Files.readString(Paths.get(rawWebhookBodyPath));
            String[] signatureParts = signature.split(":");

            String timestamp = signatureParts[0];
            // int ubbleSignatureVersion = Integer.parseInt(signatureParts[1]);
            String ubbleSignature = signatureParts[2];

            // First we create the signedPayload
            String signedPayload = timestamp + ":" + rawWebhookBody;
            byte[] signedPayloadBytes = signedPayload.getBytes("UTF-8");

            // Use the public key from the dashboard (check version from ubbleSignatureVersion)
            String publicKeyString = new String(Files.readAllBytes(Paths.get(keyPath)), Charset.defaultCharset());
            String cleanedPublicKeyString = publicKeyString
                .replace("-----BEGIN PUBLIC KEY-----", "")
                .replaceAll(System.lineSeparator(), "")
                .replace("-----END PUBLIC KEY-----", "");
            byte[] publicKeyBytes = Base64.getDecoder().decode(cleanedPublicKeyString);
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("EC");
            PublicKey publicKey = keyFactory.generatePublic(keySpec);

            // Verify the signature
            Signature verifier = Signature.getInstance("SHA512withECDSA");
            verifier.initVerify(publicKey);
            verifier.update(signedPayloadBytes);
            boolean result = verifier.verify(Base64.getDecoder().decode(ubbleSignature));
            if (result) {
                System.out.println("Success");
            } else {
                System.out.println("Could not validate signature");
            }
        } catch (IOException e) {
            System.out.println("An error occurred while reading files: " + e.getMessage());
            e.printStackTrace();
        } catch (Exception e) {
            System.out.println("An error occurred: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
